File: src/main/java/com/amazonaws/sample/cdk/discourse/CertificateNestedStack.java
package com.amazonaws.sample.cdk.discourse;

import software.amazon.awscdk.NestedStack;
import software.amazon.awscdk.services.certificatemanager.CertificateValidation;
import software.amazon.awscdk.services.certificatemanager.DnsValidatedCertificate;
import software.amazon.awscdk.services.route53.IPublicHostedZone;
import software.constructs.Construct;

public class CertificateNestedStack extends NestedStack {

    public DnsValidatedCertificate certificate;

    public CertificateNestedStack(final Construct scope, final String id, final IPublicHostedZone hostedZone, final String domainName) {
        super(scope, id);

        createCertificate(hostedZone, domainName, id);
    }

    public String getCertificateArn() {
        return certificate.getCertificateArn();
    }

    private void createCertificate(final IPublicHostedZone hostedZone, final String domainName, final String id) {
        certificate = DnsValidatedCertificate.Builder.create(this, id + "Certificate")
                .hostedZone(hostedZone)
                .domainName(domainName)
                .cleanupRoute53Records(true)
                .region(getRegion())
                .validation(CertificateValidation.fromDns(hostedZone))
                .build();
    }
}

File: lib/certificate-nested-stack.ts
import * as process from "process";
import {Construct} from 'constructs';
import {NestedStack, NestedStackProps} from "aws-cdk-lib";
import {IPublicHostedZone} from "aws-cdk-lib/aws-route53";
import {CertificateValidation, DnsValidatedCertificate} from "aws-cdk-lib/aws-certificatemanager";

interface CertificateNestedStackProps extends NestedStackProps {
    readonly hostedZone: IPublicHostedZone;
}

export class CertificateNestedStack extends NestedStack {
    certificate: DnsValidatedCertificate;

    constructor(scope: Construct, id: string, props: CertificateNestedStackProps) {
        super(scope, id, props);
        this.certificate = new DnsValidatedCertificate(this, id + "Certificate", {
            hostedZone: props.hostedZone,
            domainName: process.env.CDK_DEPLOY_DISCOURSE_DOMAIN_NAME || '',
            cleanupRoute53Records: true,
            region: this.region,
            validation: CertificateValidation.fromDns(props.hostedZone)
        });
    }
}
